src/coordination/coordinator_log_store.cpp

```diff
@@ -13,229 +13,164 @@
-    : start_idx_(1),
-      raft_server_bwd_pointer_(nullptr),
-      disk_emul_delay(0),
-      disk_emul_thread_(nullptr),
-      disk_emul_thread_stop_signal_(false),
-      disk_emul_last_durable_index_(0) {
-  // Dummy entry for index 0.
-  ptr<buffer> buf = buffer::alloc(sz_ulong);
-  logs_[0] = cs_new<log_entry>(0, buf);
-}
 
-CoordinatorLogStore::~CoordinatorLogStore() {
-  if (disk_emul_thread_) {
-    disk_emul_thread_stop_signal_ = true;
-    // disk_emul_ea_.invoke();
-    if (disk_emul_thread_->joinable()) {
-      disk_emul_thread_->join();
-    }
-  }
-}
+namespace {
 
-ptr<log_entry> CoordinatorLogStore::MakeClone(const ptr<log_entry> &entry) {
-  // NOTE:
-  //   Timestamp is used only when `replicate_log_timestamp_` option is on.
-  //   Otherwise, log store does not need to store or load it.
-  ptr<log_entry> clone = cs_new<log_entry>(entry->get_term(), buffer::clone(entry->get_buf()), entry->get_val_type(),
-                                           entry->get_timestamp());
-  return clone;
+ptr<log_entry> MakeClone(const ptr<log_entry> &entry) {
+  return cs_new<log_entry>(entry->get_term(), buffer::clone(entry->get_buf()), entry->get_val_type(),
+                           entry->get_timestamp());
 }
 
-ulong CoordinatorLogStore::next_slot() const {
-  std::lock_guard<std::mutex> l(logs_lock_);
-  // Exclude the dummy entry.
-  return start_idx_ + logs_.size() - 1;
+}  // namespace
+
+CoordinatorLogStore::CoordinatorLogStore() : start_idx_(1) {
+  ptr<buffer> buf = buffer::alloc(sizeof(uint64_t));
+  logs_[0] = cs_new<log_entry>(0, buf);
 }
 
-ulong CoordinatorLogStore::start_index() const { return start_idx_; }
+CoordinatorLogStore::~CoordinatorLogStore() {}
 
-ptr<log_entry> CoordinatorLogStore::last_entry() const {
-  ulong next_idx = next_slot();
-  std::lock_guard<std::mutex> l(logs_lock_);
-  auto entry = logs_.find(next_idx - 1);
+auto CoordinatorLogStore::FindOrDefault_(uint64_t index) const -> ptr<log_entry> {
+  auto entry = logs_.find(index);
   if (entry == logs_.end()) {
     entry = logs_.find(0);
   }
+  return entry->second;
+}
 
-  return MakeClone(entry->second);
+uint64_t CoordinatorLogStore::next_slot() const {
+  auto lock = std::lock_guard{logs_lock_};
+  return start_idx_ + logs_.size() - 1;
 }
 
-ulong CoordinatorLogStore::append(ptr<log_entry> &entry) {
-  ptr<log_entry> clone = MakeClone(entry);
+uint64_t CoordinatorLogStore::start_index() const { return start_idx_; }
 
-  std::lock_guard<std::mutex> l(logs_lock_);
-  size_t idx = start_idx_ + logs_.size() - 1;
-  logs_[idx] = clone;
+ptr<log_entry> CoordinatorLogStore::last_entry() const {
+  auto lock = std::lock_guard{logs_lock_};
 
-  if (disk_emul_delay) {
-    uint64_t cur_time = timer_helper::get_timeofday_us();
-    disk_emul_logs_being_written_[cur_time + disk_emul_delay * 1000] = idx;
-    // disk_emul_ea_.invoke();
-  }
+  uint64_t const last_idx = start_idx_ + logs_.size() - 1;
+  auto const last_src = FindOrDefault_(last_idx - 1);
```

Comment:
`uint64_t const last_idx = start_idx_ + logs_.size() - 1;` has a `-1` and `FindOrDefault_(last_idx - 1);` has is also.
Is this correct?

Reply:
yes

Reply:
because of start_idx_